auth/impersonate/idtoken.go

```diff
@@ -0,0 +1,172 @@
+	// will try to be detected from the environment. Optional.
+	TokenProvider auth.TokenProvider
+	// Client configures the underlying client used to make network requests
+	// when fetching tokens. If provided the client should provide it's own
+	// credentials at call time. Optional.
+	Client *http.Client
+}
+
+var (
+	defaultAud   = "https://iamcredentials.googleapis.com/"
+	defaultScope = "https://www.googleapis.com/auth/cloud-platform"
+)
+
+// NewIDTokenProvider creates an impersonated
+// [cloud.google.com/go/auth/TokenProvider] that returns ID tokens configured
+// with the provided config and using credentials loaded from Application
+// Default Credentials as the base credentials if not provided with the opts.
+// The tokens produced are valid for one hour and are automatically refreshed.
+func NewIDTokenProvider(opts *IDTokenOptions) (auth.TokenProvider, error) {
+	if opts == nil {
+		return nil, fmt.Errorf("impersonate: opts must be provided")
+	}
+	if opts.Audience == "" {
+		return nil, fmt.Errorf("impersonate: an audience must be provided")
+	}
+	if opts.TargetPrincipal == "" {
+		return nil, fmt.Errorf("impersonate: a target service account must be provided")
+	}
+
+	var client *http.Client
+	if opts.Client == nil && opts.TokenProvider == nil {
+		var err error
+		client, err = httptransport.NewClient(&httptransport.Options{
+			DetectOpts: &detect.Options{
+				Audience: defaultAud,
+				Scopes:   []string{defaultScope},
+			},
+		})
+		if err != nil {
+			return nil, err
+		}
+	} else if opts.Client == nil {
+		client = internal.CloneDefaultClient()
+		if err := httptransport.AddAuthorizationMiddleware(client, opts.TokenProvider); err != nil {
+			return nil, err
+		}
+	} else {
+		client = opts.Client
+	}
+
+	itp := impersonatedIDTokenProvider{
+		client:          client,
+		targetPrincipal: opts.TargetPrincipal,
+		audience:        opts.Audience,
+		includeEmail:    opts.IncludeEmail,
+	}
+	for _, v := range opts.Delegates {
+		itp.delegates = append(itp.delegates, formatIAMServiceAccountName(v))
+	}
+	return auth.NewCachedTokenProvider(itp, nil), nil
+}
+
+type generateIDTokenRequest struct {
+	Audience     string   `json:"audience"`
+	IncludeEmail bool     `json:"includeEmail"`
+	Delegates    []string `json:"delegates,omitempty"`
+}
+
+type generateIDTokenResponse struct {
+	Token string `json:"token"`
+}
+
+type impersonatedIDTokenProvider struct {
+	client *http.Client
+
+	targetPrincipal string
+	audience        string
+	includeEmail    bool
+	delegates       []string
+}
+
+func (i impersonatedIDTokenProvider) Token(ctx context.Context) (*auth.Token, error) {
+	now := time.Now()
+	genIDTokenReq := generateIDTokenRequest{
+		Audience:     i.audience,
+		IncludeEmail: i.includeEmail,
+		Delegates:    i.delegates,
+	}
+	bodyBytes, err := json.Marshal(genIDTokenReq)
+	if err != nil {
+		return nil, fmt.Errorf("impersonate: unable to marshal request: %v", err)
```

Comment:
```suggestion
		return nil, fmt.Errorf("impersonate: unable to marshal request: %w", err)
```
Maybe wrap? Here and elsewhere in this file.

Reply:
done.